src/protocol.rs

```diff
@@ -121,145 +91,56 @@ impl Participant {
-    /// is ready for the specified identifier.
-    #[instrument(skip_all)]
-    pub fn initialize_auxinfo_message(&self, auxinfo_identifier: Identifier) -> Message {
-        info!("Auxinfo generation is ready.");
-        Message::new(
-            MessageType::Auxinfo(AuxinfoMessageType::Ready),
-            auxinfo_identifier,
-            self.id,
-            self.id,
-            &[],
-        )
-    }
+        // Check that message belongs to correct protocol
+        match (message.message_type(), P::protocol_type()) {
+            (MessageType::Auxinfo(_), ProtocolType::AuxInfo)
+            | (MessageType::Keygen(_), ProtocolType::Keygen)
+            | (MessageType::Presign(_), ProtocolType::Presign) => {}
+            _ => {
+                error!(
+                    "Message type did not match type of this participant: got {:?}, expected {:?}",
+                    message.message_type(),
+                    P::protocol_type()
+                );
+                Err(CallerError::WrongProtocol)?
+            }
+        };
 
-    /// Produces a message to signal to this participant that keyshare
-    /// generation is ready for the specified identifier
-    #[instrument(skip_all)]
-    pub fn initialize_keygen_message(&self, keygen_identifier: Identifier) -> Message {
-        info!("Keyshare generation is ready.");
-        Message::new(
-            MessageType::Keygen(KeygenMessageType::Ready),
-            keygen_identifier,
-            self.id,
-            self.id,
-            &[],
-        )
+        // Handle it!
+        let outcome = self
+            .participant
+            .process_message(rng, message, &self.input)?;
+        let (output, messages) = outcome.into_parts();
+        Ok((message.id(), output, messages))
     }
 
-    /// Produces a message to signal to this participant that presignature
-    /// generation is ready for the specified identifier. This also requires
-    /// supplying the associated auxinfo identifier and keyshare identifier.
-    /// `auxinfo_identifier`, `keyshare_identifier` and `identifier` correspond
-    /// to session identifiers.
+    /// Produce a message to signal to this participant that the protocol can
+    /// begin.
     #[instrument(skip_all)]
-    pub fn initialize_presign_message(
-        &mut self,
-        auxinfo_identifier: Identifier,
-        keyshare_identifier: Identifier,
-        identifier: Identifier,
-    ) -> Result<Message> {
-        info!("Presignature generation is ready.");
-        self.presign_participant.initialize_presign_message(
-            auxinfo_identifier,
-            keyshare_identifier,
-            identifier,
-        )
+    pub fn initialize_message(&self) -> Message {
+        info!("Initializing subprotocol.");
+        Message::new(P::ready_type(), self.sid, self.id, self.id, &[])
     }
 
+    /* skipping this for now because I think it's a rabbit hole of a refactor
```

Comment:
discussion topic: I commented these out because I didn't want to figure out how to check for done-ness, or to add #160 in this PR as well, but it turns out I don't need them any more because I can just check (in a good protocol run) whether all parties returned output. 

We probably still want to do #160 so we can check bad protocol runs (or e.g. that tests fail the expected way) but is there any objection to my removing this functionality now? I can update 160 with an "add public API method to check done-ness" so we remember to bring it back.

Reply:
@amaloz Do you have an opinion on this?

Reply:
I'm all for removing this functionality now and keeping #160 around for bad protocol runs.